packages/nimble-components/src/select/index.ts

```diff
@@ -969,39 +984,89 @@ export class Select
         this.updateListboxMaxHeightCssVariable();
     }
 
+    private isOptionHidden(option: ListOption): boolean {
+        if (option.hidden) {
+            return true;
+        }
+        return !this.filterMatchesText(option.text);
+    }
+
+    private filterMatchesText(text: string): boolean {
+        const filter = this.filter.toLowerCase();
+        const normalizedFilter = diacriticInsensitiveStringNormalizer(filter);
+        return diacriticInsensitiveStringNormalizer(text).includes(
+            normalizedFilter
+        );
+    }
+
     /**
      * Filter available options by text value.
      *
      * @public
      */
     private filterOptions(): void {
-        const filter = this.filter.toLowerCase();
-
-        if (filter) {
-            this.filteredOptions = this.options.filter(option => {
-                const normalizedFilter = diacriticInsensitiveStringNormalizer(filter);
-                return (
-                    !option.hidden
-                    && diacriticInsensitiveStringNormalizer(option.text).includes(
-                        normalizedFilter
-                    )
-                );
-            });
-        } else {
-            this.filteredOptions = this.options.filter(
-                option => !option.hidden
-            );
+        if (!this.$fastController.isConnected) {
+            return;
         }
 
-        this.options.forEach(o => {
-            if (isNimbleListOption(o)) {
-                if (!this.filteredOptions.includes(o)) {
-                    o.visuallyHidden = true;
-                } else {
-                    o.visuallyHidden = false;
+        const filteredOptions: ListOption[] = [];
+
+        let lastVisibleOptionGroupIndex = -1;
+        let lastVisibleElement: ListOption | ListOptionGroup | undefined;
+        for (let i = 0; i < this.slottedOptions.length; i++) {
+            const element = this.slottedOptions[i]!;
+
+            if (element instanceof ListOptionGroup) {
+                element.showTopSeparator = false;
+                element.showBottomSeparator = false;
```

Comment:
Do you need to reset these to `false`? They get set if the option is visible, and I think it would be good if the styling was written such that the separators aren't shown on a hidden group, regardless of the values of these properties.